File: verify_serializer_tmp.py

class ListNode:
    def __init__(self, val, next=None):
        self.val = val
        self.next = next

def list_node_to_json(head):
    if not head: return None
    result = []
    visited = set()
    curr = head
    while curr and curr not in visited:
        visited.add(curr)
        result.append(curr.val)
        curr = curr.next
    return result

# Test cases
node3 = ListNode(3)
node2 = ListNode(2, node3)
head = ListNode(1, node2)

print(f"Normal list: {list_node_to_json(head)}")

node3.next = node2 # Cycle
print(f"Cyclic list: {list_node_to_json(head)}")

print(f"Empty list: {list_node_to_json(None)}")
print(f"Single node: {list_node_to_json(ListNode(5))}")

File: verify_serializer_tmp.js

class ListNode {
    constructor(val, next = null) {
        this.val = val;
        this.next = next;
    }
}

function listNodeToJson(head) {
    if (!head) return null;
    const result = [];
    const visited = new Set();
    let curr = head;
    while (curr && !visited.has(curr)) {
        visited.add(curr);
        result.push(curr.val);
        curr = curr.next;
    }
    return result;
}

// Test cases
const node3 = new ListNode(3);
const node2 = new ListNode(2, node3);
const head = new ListNode(1, node2);

console.log("Normal list:", JSON.stringify(listNodeToJson(head))); // [1, 2, 3]

node3.next = node2; // Cycle
console.log("Cyclic list:", JSON.stringify(listNodeToJson(head))); // [1, 2, 3]

console.log("Empty list:", JSON.stringify(listNodeToJson(null))); // null
console.log("Single node:", JSON.stringify(listNodeToJson(new ListNode(5)))); // [5]
